### ISSUE TITLE
MSVC 2019 (Windows 10): debug assertion in xmemory on spirv_cross::Compiler destruction

### ISSUE THREAD

**Issue by goulash32** [KhronosGroup/SPIRV-Cross#1438]
I'm using SPIRV-Cross, statically linked (spirv-cross-cored.lib) in my MSVC 2019 64-bit debug Vulkan project, built using the same compiler with /MTd (multithreaded-debug) specified--same as my project. I'm strictly using the reflection capabilities of the library and not doing cross-compilation.

My SPIRV code is compiled from valid Vulkan-flavored GLSL using the shaderc library, and there are no errors or warnings reported by that library during compilation.

I create a new spirv_cross::Compiler instance and pass it my SPIRV code, which works without issue (using either the vector-based or ptr-based constructors). I am then able to get and iterate over all my shader resources, again without issue.

The problem arises when the SPIRV-Cross compiler goes out of scope and its destructor is called. I receive a debug assertion failure (see attached image). This issue also arises if I simply create the Compiler instance and then destroy it without using it. I have also attached the call stack here, which hopefully sheds some light on exactly where this exception occurs. The comment on the offending assertion in the xmemory source is _"If the following asserts, it likely means that we are performing an aligned delete on memory coming from an unaligned allocation."_

I can't find any other examples of this issue in the community. Any insight would be greatly appreciated!

![debugassertion](https://user-images.githubusercontent.com/6015156/88963561-dde7c680-d25c-11ea-8f53-b8c9faa87d4b.png)

![callstack](https://user-images.githubusercontent.com/6015156/88963567-e0e2b700-d25c-11ea-8a68-712c679bd2e0.png)

Offending block:
```
std::unique_ptr<ShaderModule> ShaderFactory::create_shader( 
	std::string filename, 
	ShaderModule::ShaderStage shader_stage 
) {
	auto shader_bytes = util::readShaderFile( filename );
	auto shader_stage_conv = vk_to_shaderc.find( shader_stage );

	if (shader_stage_conv == vk_to_shaderc.end()) {
		throw std::runtime_error( "Invalid Shader Stage specified during compilation." );
	}

	auto shader_kind = shader_stage_conv->second;

	shaderc::SpvCompilationResult result;
	try {
		shaderc::CompileOptions opt;
		opt.SetTargetEnvironment( shaderc_target_env::shaderc_target_env_vulkan, shaderc_env_version_vulkan_1_2 );
				
		result = compiler->CompileGlslToSpv( 
			shader_bytes.data(), 
			shader_bytes.size(), 
			shader_kind, 
			filename.c_str(), 
			opt 
		);

		if (result.GetCompilationStatus() != shaderc_compilation_status::shaderc_compilation_status_success) {
			throw std::runtime_error( result.GetErrorMessage() );
		}
	} catch (std::exception e) {
		std::stringstream ss;
		ss << "Unable to compile Shader: " << e.what();
		throw std::runtime_error( ss.str() );
	}

	size_t code_length = result.end() - result.begin();
	const uint32_t* code = result.begin();

	std::vector<uint32_t> code_vec( code, code + code_length );
	spirv_cross::Compiler compiler( std::move( code_vec ) );

	auto resources = compiler.get_shader_resources();
	auto built_resources = build_shader_resources( shader_stage, compiler, resources );
	size_t code_size = code_length * sizeof( uint32_t );
	auto shader_module = std::make_unique<ShaderModule>(
		parent_device,
		code,
		static_cast<uint32_t>(code_size),
		shader_stage,
		std::move( built_resources )
	);

	return shader_module;
// Exception occurs when the spirv_cross::Compiler goes out of scope here.
}
```

**Comment by Firnor** [comment]
I got the exact same issue on my CMake project when I built it with MSVC.

However, linking to all the spirv libraries (except msl and hlsl) fixed it for me.

**Comment by goulash32** [comment]
Thanks for the suggestion Firnor. I tried linking all combinations of the generated .libs to no avail. However, I had been building the library using the generated .sln in Visual Studio rather than using CMake. This was the source of the problem.

I had to add the following two lines to CMakeLists.txt:

```
cmake_policy(SET CMP0091 NEW)
set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```

in order to get a MultiThreaded(Debug) (as opposed to MultiThreaded(Debug)DLL). After doing so, I did a CMake build, linked only spirv-cross-core(d).lib and spirv-cross-cpp(d).lib and the crash stopped.
